globalsat_sport.cc

```diff
@@ -0,0 +1,742 @@
+/*
+    Globalsat Sport serial protocol.
+
+    Copyright (C) 2012, Zingo Andersen
+
+    Dual Licese Public Domain and GPLv2 or later
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA
+
+ */
+
+/*
+ * This is the bridge between the GPSBable and globalsport device
+ * gh625XT
+ *
+ *
+ * ./gpsbabel -i globalsat -f /dev/ttyUSB0 -o gpx,garminextensions -F xxx.gpx
+ * ./gpsbabel -i globalsat -f /dev/ttyUSB0 -o gtrnctr,course=0,sport=Running -F xxx.fit
+ *
+ * ./gpsbabel-debug -i globalsat -f /dev/ttyUSB0 -o gpx,garminextensions -F xxx.gpx
+ */
+
+
+#include <ctype.h>
+#include <limits.h>
+#include <endian.h>
+#include <stdio.h>
+#include "defs.h"
+#include "gbser.h"
+
+#define MYNAME "GlobalsatSport"
+
+static void *serial_handle;
+static int isSizeSwaped;
+
+static char *showlist = NULL;    //if true show a list instead of download tracks
+static char *track = 0;    //if not 0 only download this track, if 0 download all
+
+static 
+arglist_t globalsat_args[] = {
+		{"showlist", &showlist, "list tracks", NULL, ARGTYPE_BOOL, ARG_NOMINMAX},
+		{"track", &track, "get track", "0", ARGTYPE_INT, ARG_NOMINMAX}, 
+		ARG_TERMINATOR
+};
+
+typedef enum
+{
+	CommandWhoAmI = 0xBF,
+	CommandGetSystemInformation = 0x85,
+	CommandGetSystemConfiguration = 0x86,
+
+	CommandSetSystemConfiguration = 0x96,
+	CommandSetSystemInformation = 0x98,
+	CommandGetScreenshot = 0x83,
+
+	CommandGetWaypoints = 0x77,
+	CommandSendWaypoint = 0x76,
+	CommandDeleteWaypoints = 0x75,
+
+	CommandSendRoute = 0x93,
+	CommandDeleteAllRoutes = 0x97,
+
+	CommandGetTrackFileHeaders = 0x78,
+	CommandGetTrackFileSections = 0x80,
+	CommandGetNextTrackSection = 0x81,
+	//CommandReGetLastSection = 0x82,
+	CommandId_FINISH = 0x8A,
+	CommandSendTrackStart = 0x90,
+	CommandSendTrackSection = 0x91,
+
+	HeaderTypeLaps = 0xAA,
+	HeaderTypeTrackPoints = 0x55,
+
+	ResponseInsuficientMemory = 0x95,
+	ResponseResendTrackSection = 0x92,
+	ResponseSendTrackFinish = 0x9A
+} globalsat_commands_e;
+
+
+//Pack structures since they should conform to serial data spec
+#pragma pack(1)
```

Comment:
We've avoided #pragma pack so far, manually packing structures when we have to just because this is such a messy area.  In fact, we usually avoid structures in protocol definitions, but I could be convinced that's a battle that's not worth fighting.